solvers/bcd_celer.py

```diff
@@ -0,0 +1,363 @@
+                   tol=1e-12, p0=10, verbose=0, prune=True, accel=True):
+    n_samples, n_features = X.shape
+    n_tasks = Y.shape[1]
+    n_obs = n_samples * n_tasks
+
+    W = np.zeros((n_features, n_tasks))
+    R = Y.copy()
+    verbose_in = max(verbose - 1, 0)
+    n_screened = 0
+    norm_Y2 = norm(Y, 'fro') ** 2
+
+    tol *= norm_Y2
+    if p0 > n_features:
+        p0 = n_features
+
+    prios = np.empty(n_features, dtype=X.dtype)
+    screened = np.zeros(n_features, dtype=np.int32)
+    notin_WS = np.zeros(n_features, dtype=np.int32)
+
+    # acceleration variables
+    K = 6
+    last_K_R = np.empty((K, n_samples, n_tasks), dtype=X.dtype)
+    U = np.empty((K - 1, n_obs), dtype=X.dtype)
+    UtU = np.empty((K - 1, K - 1), dtype=X.dtype)
+
+    norms_X_block = norm(X, axis=0)
+    inv_lc = 1 / norms_X_block ** 2
+
+    gaps = np.zeros(n_iter, dtype=X.dtype)
+
+    Theta = np.zeros((n_samples, n_tasks), dtype=X.dtype)
+    Theta_in = np.zeros((n_samples, n_tasks), dtype=X.dtype)
+    Thetacc = np.zeros((n_samples, n_tasks), dtype=X.dtype)
+
+    # d_obj_from_inner = 0.
+    all_features = np.arange(n_features)
+    C = all_features.copy()
+    ws_size = p0
+
+    for t in range(n_iter):
+        Theta[:] = R / alpha
+
+        norm_XT_Theta = norm(X.T @ Theta, axis=1)
+        scal = max(norm_XT_Theta)
+        # The piece of code below is slow.
+        # scal, norm_XT_Theta = dual_scaling_mtl(
+        # Theta, X, all_features, screened)
+
+        if scal > 1.:
+            Theta /= scal
+            norm_XT_Theta /= scal
+        d_obj = dual_mtl(alpha, norm_Y2, Theta, Y)
+
+        # The piece of code below is slow.
+        # if t > 0:
+        #     scal, norm_XT_theta_in = dual_scaling_mtl(
+        #         Theta_in, X, all_features, screened)
+        #     if scal > 1.:
+        #         Theta_in /= scal
+        #         norm_XT_theta_in /= scal
+        # d_obj_from_inner = dual_mtl(alpha, norm_Y2, Theta_in, Y)
+        # if d_obj_from_inner > d_obj:
+        #     d_obj = d_obj_from_inner
+        #     Theta[:] = Theta_in
+        #     norm_XT_Theta[:] = norm_XT_theta_in
+
+        highest_d_obj = d_obj
+
+        p_obj = primal_mtl(W, alpha, R)
+        gap = p_obj - highest_d_obj
+        gaps[t] = gap
+
+        if verbose:
+            print("Iter {:d}: primal {:.10f}, gap {:.2e}".format(
+                  t, p_obj, gap), end="")
+
+        if gap <= tol:
+            if verbose:
+                print("\nEarly exit, gap: {:.2e} < {:.2e}".format(gap, tol))
+            break
+
+        radius = np.sqrt(2 * gap) / alpha
+
+        n_screened = set_prios_mtl(X, W, norms_X_block, prios, screened,
+                                   radius, n_screened, norm_XT_Theta)
+        ws_size = create_ws_mtl(prune, W, prios, p0, t, screened, C,
+                                n_screened, ws_size)
+        # if ws_size == n_features then argpartition will break
+        if ws_size == n_features:
```

Comment:
this is cython inherited stuff where I was lost when I wrote it, IMO you can use a full np.sort and take the highest values  (who cares to sort a p array instead of just getting the WS_size highest values), so as to avoid these 5 lines which are confusing. 

There should be multiple occurrences of this block whih can be simplified

Reply:
I found only one.

Reply:
there is this one and bcd_celer_free, no ?

Reply:
(also worth remembering for the Lasso benchmark)

Reply:
I removed bcd_celer_free to have and only solver that handles free and fixed orient, so that it does not create additional confusion.